testing/E2E_Test_Update.md

```diff
@@ -0,0 +1,164 @@
+to cover all platforms that need to be covered if abstractions such as page objects and 
+data objects are used to remove selectors and expected data from the the tests themselves.  
+These abstractions can then be passed into the tests for multiple platforms.  Indeed, the 
+Java-client has support for cross-platform page objects through the use of special 
+annotations.
+
+## Updated Proposals
+
+A decision should first be made if preferred selectors are to be implmented in the mobile
+showcase applications.  The effects of this decision are shown below:
+
+#### 1). Using Preferred Selectors in Applications
+
+* Tests are likely to be less brittle, therefore less time should be required on 
+maintenance
+* Any client can be used provided it has the ability to handle redirects to the browser
+
+
+#### 2). Not Using Preferred Selectors in Applications
+
+* Tests are likely to be more brittle, therefore more time required to maintain them
+* The Java-client should be the preferred choice of client as it has more strategies to
+target locators than other libraries
+
+### Possible Test Suite Implementations
+
+1. A nodejs implmentation with webdriverio - This should be considered if there is a 
+decision to use preferred selecotrs is to use them in the showcase applications as there 
+is a lot of experience using nodejs within the teams.
+
+1. A Java-client implementation or the Akow library - should be considered if the decision
+on using preferred selectors is not to use them. This will allow for more locator strategies 
+that can be used to create the tests.  The downside of this approach could be the upkeep of 
+the Akow library. 
+
+## Test Writing Time Estimation
+
+### Writing New tests
+
+A new test should take approximately 3-4 hours to complete regardless of the language of 
+implementation.  Appium Inspector can be used to find the selectors and also generate an 
+initial version of the code that is unrefined.  The person developing the test will need 
+to extract the selectors and generate page objects and test case files based on the 
+unrefined code.  
+
+Through the use of best practices such as page objects, areas of a test that have been 
+previously covered by code inside the page object can be quickly reused in the new test 
+cases and the only selectors and code to be created will belong to pages that have yet 
+to be covered by a test case.
+
+However, in cases where preferred selector strategies cannot be used or in cases where 
+there may be some problems with the application some extra time may be required.
+
+
+### Maintaining tests
+
+The majority of time spent to make changes to maintain the test suite should be minimal 
+(1 – 2 hours).  
+
+If the changes are made to the UI that affect the UI selectors in the application the 
+selectors in the page objects should be updated to reflect them and this should take the 
+shortest amount of time.  
```

Comment:
If we have test in for example showcase app. It can be included in CI solution, so author of the UI change sees that it fails immediately. But I'm not sure that Appium is that stable (for iOS especially) that we can always rely on the test results.

Reply:
If you are using TDD then the author should edit test first before making UI changes, if not then test suite maintainer.

Currently, there is no Swift client for Appium.

Placing these UI tests inside the showcase apps means that you are going to be recreating the same tests, framework initiation and data on four different platforms using four different languages, which gives you a lot more potential "points of failure" or "points of divergence" within the tests.

An environment file could be placed within the test suite that points to the location of where the example apps binary is produced for each platform, allowing the developer to run changes locally as well as through CI.

The advantages of a single test suite are:
 - one language (although Kotlin's interoperability with Java means you could write page objects as POJOs)
- one data set
- one set of tests (where you need differences in the flow of applications for one reason or another you can use some form of flow control n the test, although some form of anchor would be required on the app's homepage to detect which example app is running (Xamarin, Swift, etc) as Appium can only detect the basic platform the app is running in (iOS or Android))
- The only areas you might have differing files is the page objects themselves